promql/engine.go

```diff
@@ -3419,6 +3478,65 @@ func makeInt64Pointer(val int64) *int64 {
 	return valp
 }
 
+// Add RatioSampler interface to allow unit-testing (previously: Randomizer).
+type RatioSampler interface {
+	// Return this sample "offset" between [0.0, 1.0]
+	sampleOffset(ts int64, sample *Sample) float64
+	AddRatioSample(r float64, sample *Sample) bool
+}
+
+// Use Hash(labels.String()) / maxUint64 as a "deterministic"
+// value in [0.0, 1.0].
+type HashRatioSampler struct{}
+
+var ratiosampler RatioSampler = NewHashRatioSampler()
+
+func NewHashRatioSampler() *HashRatioSampler {
+	return &HashRatioSampler{}
+}
+
+func (s *HashRatioSampler) sampleOffset(ts int64, sample *Sample) float64 {
+	const (
+		float64MaxUint64 = float64(math.MaxUint64)
```

Comment:
I guess this is ok in practice, although it makes me at least slightly uncomfortable that a float64 won't be able to represent anything above 2^53 without loss of precision (so you won't store exactly `math.MaxUint64` here, but only approximately).

Reply:
Same precision "loss" is present at the final
```
return float64(sample.Metric.Hash()) / float64MaxUint64
```
Given that strict number precision is not needed here, especially given the ~unpredictable  nature of `Metric.Hash() uint64` also (_unpredictable_ from the _PoV_ of the user who doesn't interact with its values), IMO it would not be an issue unless `ratio(1.0, metric)` wouldn't return all samples (verified with some `testdata/`).

Reply:
Yes, that's what I also roughly arrived at - thanks for confirming :)

Reply:
cool yw+